### PR TITLE
refactor: refresh token rotation

### PR DESCRIPTION
**Pull request by aeneasr**
## Checklist

- [ ] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md) and signed the CLA.
- [ ] I have referenced an issue containing the design document if my change introduces a new feature.
- [ ] I have read the [security policy](../security/policy).
- [ ] I confirm that this pull request does not address a security vulnerability. 
      If this pull request addresses a security vulnerability, 
      I confirm that I got approval (please contact [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push the changes.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] I have added the necessary documentation within the code base (if appropriate).

### CODE REVIEW

**Review by alnr — APPROVED** [review]

File: handler/oauth2/flow_refresh.go [hunk 1]
```diff
@@ -51,25 +51,31 @@ func (c *RefreshTokenGrantHandler) HandleTokenEndpointRequest(ctx context.Contex
 			return errorsx.WithStack(rErr)
 		}
 
-		return errorsx.WithStack(fosite.ErrInactiveToken.WithWrap(err).WithDebug(err.Error()))
+		return fosite.ErrInvalidGrant.WithWrap(err).
```
Comment (zepatrik) [inline handler/oauth2/flow_refresh.go:54]: Why remove the stack for all these errors?
Also, why change the explicit error here? Could this be breaking? From the hint it seems like this should still be `fosite.ErrInactiveToken`.

File: handler/oauth2/flow_refresh.go [hunk 2]
```diff
@@ -51,25 +51,31 @@ func (c *RefreshTokenGrantHandler) HandleTokenEndpointRequest(ctx context.Contex
...
+			WithDebugf("Refresh token re-use was detected. All related tokens have been revoked.")
 	} else if errors.Is(err, fosite.ErrNotFound) {
-		return errorsx.WithStack(fosite.ErrInvalidGrant.WithWrap(err).WithDebugf("The refresh token has not been found: %s", err.Error()))
+		return fosite.ErrInvalidGrant.WithWrap(err).
+			WithHint("The refresh token is malformed or not valid.").
+			WithDebug("The refresh token can not be found.")
 	} else if err != nil {
-		return errorsx.WithStack(fosite.ErrServerError.WithWrap(err).WithDebug(err.Error()))
-	} else if err := c.RefreshTokenStrategy.ValidateRefreshToken(ctx, originalRequest, refresh); err != nil {
+		return fosite.ErrInvalidGrant.WithWrap(err).WithDebug(err.Error())
```
Comment (zepatrik) [inline handler/oauth2/flow_refresh.go:62]: Also here, the error should probably stay `fosite.ErrServerError`.

**Review by zepatrik — CHANGES_REQUESTED** [review]

File: handler/oauth2/flow_refresh.go (hunk 1 above)
Comment (zepatrik) [reply handler/oauth2/flow_refresh.go:54]: OK, the spec is just weird here and `fosite.ErrInvalidGrant` is correct.

Comment (aeneasr) [reply handler/oauth2/flow_refresh.go:54]: > Why remove the stack for all these errors?

`WithWrap` uses the passed error to do the stack trace - so the internal error already has a stack trace and then we just re-use this. It's a bit like `fmt.Errorf("%w",err)`

File: handler/oauth2/flow_refresh.go [hunk 3]
```diff
@@ -130,30 +136,20 @@ func (c *RefreshTokenGrantHandler) PopulateTokenEndpointResponse(ctx context.Con
...
-		return err
-	} else if err := c.TokenRevocationStorage.RevokeAccessToken(ctx, ts.GetID()); err != nil {
-		return err
-	}
+	storeReq := requester.Sanitize([]string{})
+	storeReq.SetID(requester.GetID())
 
-	if err := c.TokenRevocationStorage.RevokeRefreshTokenMaybeGracePeriod(ctx, ts.GetID(), signature); err != nil {
-		return err
+	if err = c.TokenRevocationStorage.RotateRefreshToken(ctx, requester.GetID(), signature); err != nil {
```
Comment (mitar) [inline handler/oauth2/flow_refresh.go:143]: @aeneasr I think this introduced a bug. After we upgraded our systems to the version with this change, no refresh tokens in our tests (where we use memory storage) got invalidated. The reason I think is that `requester` here has a completely new and random ID, which of course has nothing with the existing refresh token.

Previous code had it right, it retrieved refresh token requester with `GetRefreshTokenSession` and used `ts.GetID()` here. I think this should be fixed and correct requester ID should be passed here.

Comment (mitar) [reply handler/oauth2/flow_refresh.go:143]: My current workaround is to fix this at store level:

```
func (s *MemoryStore) RotateRefreshToken(ctx context.Context, _ string, refreshTokenSignature string) error {
	ts, err := s.GetRefreshTokenSession(ctx, refreshTokenSignature, nil)
	if err != nil {
		return err
	}
	if err := s.RevokeRefreshToken(ctx, ts.GetID()); err != nil {
		return err
	}
	return s.RevokeAccessToken(ctx, ts.GetID())
}
```

But this is just a workaround, I think, it is completely useless to pass new requester ID into this function. Or is this something you use and need in Hydra?

Comment (aeneasr) [reply handler/oauth2/flow_refresh.go:143]: The request ID is supposed to remain stable for the lifetime duration of a token chain, so this is correct! You can check the implementation in Hydra for reference

Comment (mitar) [reply handler/oauth2/flow_refresh.go:143]: You are right. It works. I had in our code a part where I was unconditionally setting request ID to a new value. The reason is that I want a bit shorter ID representation than what default `uuid.New().String()` provides. I will have to find another way to do that. I could not find an easy way to patch `GetID()` to use a different ID constructor. :-(

Thanks.

Comment (aeneasr) [reply handler/oauth2/flow_refresh.go:143]: Fosite’s not the easiest library to work with

Comment (mitar) [reply handler/oauth2/flow_refresh.go:143]: Oh, it is great!

**Pull request merged**